frappe/model/base_document.py

```diff
@@ -71,15 +71,19 @@ def _get_controller():
 
 	return site_controllers[doctype]
 
+# This is default value constant for performance improvement in BaseDocument methods.
+_DEFAULT_VAL = object()
```

Comment:
Requires better reasoning and references. By itself, it seems unintuitive and counterproductive.

Reply:
About this practice: https://www.google.com/search?hl=en&q=unique%20custom%20sentinel%20object%20python

This helps to reduce `hasattr()` and `getattr()` calls.

For example, we could do:
```python
if key in dict_:
    self.set(key, value)
```
or
```python
val = dict_.get(key, _DEFAULT_VAL)
if val is not _DEFAULT_VAL:
    self.set(key, value)
```

Reply:
Actually, it is slower. I will revert the change.